File: Data Structures & Algorithms/combination-target-sum/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @returns {number[][]}
     */
    combinationSum(nums, target) {
        const dfs = (acc, cur, sum, i) => {
            if (i >= nums.length || sum > target) return acc
            if (sum === target) return [...acc, cur]

            const left = dfs(acc, [...cur, nums[i]], sum + nums[i], i)
            const right = dfs(left, [...cur], sum, i + 1)
            return right
        }

        const result = dfs([], [], 0, 0)
        return result
    }
}

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-0.py
from functools import lru_cache

class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        full = sum(nums)
        if full % 2 == 1:
            return False
        
        half = full // 2
        n = len(nums)

        @lru_cache(maxsize=None)
        def dfs(i: int, target: int) -> bool:
            if target == 0:
                return True
            if i == n or target < 0:
                return False

            return dfs(i + 1, target - nums[i]) or dfs(i + 1, target)

        return dfs(0, half)